data/Builder.go

```diff
@@ -0,0 +1,72 @@
+package data
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/tidepool-org/platform/validate"
+)
+
+type Datum map[string]interface{}
+
+type DatumArray []Datum
+
+type Builder interface {
+	BuildFromDatum(datum Datum) interface{}
+	BuildFromDatumArray(datumArray DatumArray) ([]interface{}, *validate.ErrorsArray)
+}
+
+type TypeBuilder struct {
+	inject map[string]interface{}
+	Errors *validate.ErrorsArray
+	Index  int
+}
+
+func NewTypeBuilder(inject map[string]interface{}) Builder {
+	return &TypeBuilder{
+		inject: inject,
+		Errors: validate.NewErrorsArray(),
+		Index:  0,
+	}
+}
+
+func (t *TypeBuilder) BuildFromDatumArray(datumArray DatumArray) ([]interface{}, *validate.ErrorsArray) {
+
+	var set []interface{}
+
+	for i := range datumArray {
+		if item := t.BuildFromDatum(datumArray[i]); item != nil {
+			set = append(set, item)
+		}
+		t.Index++
+	}
+	if t.Errors.HasErrors() {
+		return nil, t.Errors
+	}
+
+	return set, nil
+}
+
+func (t *TypeBuilder) BuildFromDatum(datum Datum) interface{} {
+
+	const typeField = "type"
+
+	if datum[typeField] != nil {
+
+		for k, v := range t.inject {
+			datum[k] = v
+		}
+
+		if strings.ToLower(datum[typeField].(string)) == strings.ToLower(BasalName) {
```

Comment:
Are we going to stick with a big `if` statement?  Or move to a "registered" factory pattern?  If the `if` statement, I suggest using a `switch` statement instead.

Also, I'm going to refine my stance on "lenient on input, strict on output". I think "keywords" (the `type` value in this case) HAS to be case-sensitive.  There's just too much potential for error by allowing case-insensitivity of everything.   If we don't fail on case issues early on, we either need to 1) force them to one case before storing; or 2) lowercase ANY check EVERYWHERE. Since #2 is really easy to forget and error prone, and #1 technically changes the data from what the client sent, I recommend just requiring the case to be correct.